### PR TITLE
Update .d.ts definitions

### PR DESCRIPTION
**Pull request by Mingun**
Working on the new way of reporting errors I realized, that `peg.d.ts` is not true. It contains several definitions that was never exist and has a lack of types, that I plan update in my new PR.

So I decided to fix definitions first. This lead to some uncompatible changes in the `GrammarError` definition, but, judging that fact that it was incorrect, nobody could use it. So we can do that uncompatible change, because that is a bug fix

### CODE REVIEW

**Comment by hildjj** [comment]
I'm happy to take this as a stopgap, but we have a working typescript pass in the build now, which generates `.d.ts` files then throws them away.  I'd like to turn on type checking, then use the output of typescript if we can get it compatible-enough with what we've had in the past.

**Comment by hildjj** [comment]
We also need a way to test this file, at least for syntax.  Maybe we should add a .ts file to the examples directory?

**Comment by Mingun** [comment]
I doubt that automatic inferring could infer that:
https://github.com/peggyjs/peggy/blob/f5fca4e00257f2350bda201690bde7fabb234215/lib/peg.d.ts#L478-L497

That code proves that you can do only that things:
```ts
let visit = visitor.build({
  simple_and() { return 1;  },
  simple_not() { return ""; },

  // no error, but this is because that method can be used by other methods
  not_exist1() { return false; },
});
let res1 = visit({ type: "simple_and" });// res1 type is a number - inferred from "simple_and"
let res2 = visit({ type: "simple_not" });// res2 type is a string - inferred from "simple_not"
let res3 = visit({ type: "not_exist2" });// error: unknown node type
```

Simplified example on [typescript playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgHIHsAmEAqBPABwgGdkBvAKGWWOAFsCAbCAfThEwH4AKELCAFzIAgsTAZsAHgBEtBszYdpAPgA0yAHRa4UAObEh7PAG0AugEpDIPAG4qNek1Z8wPPtiGjx-GXKcsXFXUtDR19KxMLCLsAXwoKUEhYRBQvCQhJHGVyezBCQWQcWITwaHgkZAA1YFowdChJADFkCAAPSA5SdPwiYmzKakyW9ohO5ABrCDx0GDR+HpJlXn5PMXTMtU1tPQNkIzNLZAAlCDAAVygQBabjHFNlYpgzkAQwYHQQZAAjM+BGTCaww6mC683yfW4Txebw+u0ah2qtXqTX69igpwunyhr3eIGWHhEax8k2ms0amxCYWI5hy1GozDAyGxyAAvEznjjYcZ3BANHkiKY7HTkMBZtwAITYmkDYXIMAACyg6AA7sgQBBVQBRKBKqDcAAGABIyDy+fkYiLPs9xnxlSB9eYhXS4sL0edLuyXviIMFtvpHfY4nEKAzkAA3GrARlsn5-TDcGV+BTsePS5BuzHIACMNmoMVU9iTznQYG4aYzHuk0lz+fi1AA9PW1egWjr6uofoyFTURaQvhAEHAzsQUAq4Iy6Kd5VhkIPPv3kMOIJhvnhkCX5dBkJOFVhiPYXCw2jUwFmy+R0xiPQAGGsFmIB0Po4hZ1nhyOlshy-JCWSOZNKMgD42I2l4vt+RC9mqZx0P2UAhqcYEAExvhGtQJhBBR-vIxZgNIQGOqBz4ofyKA9mIUCgLoCGMs+ADMqEfhhpG-oex5iEh+HAaB0C6kI1q2p8PKYUAA) (strange, but it reports errors that my `tsc` (version=4.2.4) do not complained. Easy fixable, although)

----
I performing tests by changing `tsconfig.json` and by checking what SonarLint VSCode plugin shows me as types of imported peggy library (I imagine that I'm write a plugin for peggy):
```diff
 tsconfig.json | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/tsconfig.json b/tsconfig.json
index fe8a4420..3f17e71e 100644
--- a/tsconfig.json
+++ b/tsconfig.json
@@ -4,7 +4,7 @@
     "module": "commonjs",                           /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', 'es2020', or 'ESNext'. */
     // "lib": [],                                   /* Specify library files to be included in the compilation. */
     "allowJs": true,                                /* Allow javascript files to be compiled. */
-    "checkJs": false,                               /* Report errors in .js files. */
+    "checkJs": true,                                /* Report errors in .js files. */
     // "jsx": "preserve",                           /* Specify JSX code generation: 'preserve', 'react-native', 'react', 'react-jsx' or 'react-jsxdev'. */
     "declaration": true,                            /* Generates corresponding '.d.ts' file. */
     "declarationMap": false,                         /* Generates a sourcemap for each corresponding '.d.ts' file. */
@@ -14,8 +14,8 @@
     "removeComments": true,                         /* Do not emit comments to output. */
 
   /* over time we will enable these */
-    // "strict": true,                              /* Enable all strict type-checking options. */
-    // "noImplicitAny": true,                       /* Raise error on expressions and declarations with an implied 'any' type. */
+    "strict": true,                                 /* Enable all strict type-checking options. */
+    "noImplicitAny": false,                         /* Raise error on expressions and declarations with an implied 'any' type. */
     // "strictNullChecks": true,                    /* Enable strict null checks. */
     // "strictFunctionTypes": true,                 /* Enable strict checking of function types. */
     // "strictBindCallApply": true,                 /* Enable strict 'bind', 'call', and 'apply' methods on functions. */
@@ -34,5 +34,5 @@
     "skipLibCheck": true,                           /* Skip type checking of declaration files. */
     "forceConsistentCasingInFileNames": true        /* Disallow inconsistently-cased references to the same file. */
   },
-  "include": ["./lib/**/*.js", "./test/**/*.spec.js", "./benchmark/*.js"]
+  "include": ["./lib/**/*.js"]
 }
```
Making tests will be a good idea, but I still have no idea where to start. Opinions?

**Comment by phpnode** [comment]
[Updated TS Playground link which I think does what you want](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgHIHsAmEAqBPABwgGdkBvAKGWWOAFsCAbCAfThEwH4AKELCAFzIAgsTAZsAHgBEtBszYdpAPgA0yAHRa4UAObEh7PAG0AugEpDIPAG4qNek1Z8wPPtiGjx-GXKcsXFXUtDR19KxMLCLsAXwoKMEIUYWsAMQBXEAQwYHQQZABeZG4QsINkIzNzQuUK6zsKUEhYRGSxCQhJHFrKakSiIRxYxvBoeCRkADVgWjB0KElU5AgAD0gOUg78ImIe+y7ltYgN5ABrCDx0GGRU5V5+T3afbuDtPXLKqOQAJQgwdKgIG2nVSxhwpmQADIRGlMtlciBlMMYHCcnlkAAjdLARiYRaHdaYTb8YG7bgorJokDlVKWKYzYBzBa3cj2KB-AH5CnwvL3DwiJ5Sc6Xa63V6hd7VXrUZDMMDIbmFBWohHEYzuCAafoQUx2GXIYDXbjaq7KrLIACEBSKAHJuVSbVL7PqwAALKDoADuyBAEG9AFEoB6oNwAAYAEjIGq1SRiBvymVOfE9IFD5j1Mri+vZ-0BZoQfIg4rK6fscTiFDlyAAbgz5UUsTjMNxpX4FOxm1LkDnOcgAIw2agxVT2NvOdBgbhdnt56TSQfD+LUAD0y596GWQfm6ix8rdMwNpAxEAQcHSxBQbrg8rof1dWGQp-yx+Q54gmExeGQE9d0GQt7dLBiHsFwWFWGYwD7KdyG7Dk8wABgXEcYlLKt2WIPslVrWYW2QbUhFkRx2yUZAUJsVdYIwvCkkPH10joY8oErP5KIAJiwutcPw5BCPkccwGkUj0wo9D2O1WixCgUBdGY+V0IAZg4nCyGogYeNA8CxFYwSyIo6BgyERNk3yDVVIgIA)

**Comment by hildjj** [comment]
Is this next?  or #144?

**Comment by Mingun** [comment]
I think that this one is better to merge first

**Comment by Mingun** [comment]
I've rebased, fix one minor thing from the previous PR ([the first commit](https://github.com/peggyjs/peggy/pull/143/commits/0e778698907eb31f6d9fe476ebdaecd3ca676455)) so final review is expected.

**Comment by hildjj** [comment]
I'm adding a few tests using [dts-jest](https://github.com/ikatyang/dts-jest) that can land after this.  One thing I've found so far is that `generate('string')` returns `any` instead of a `Parser`

File: lib/peg.d.ts [hunk 1]
```diff
@@ -1,42 +1,224 @@
 // Based on PEG.js Type Definitions by: vvakame <https://github.com/vvakame>, Tobias Kahlert <https://github.com/SrTobi>, C.J. Bell <https://github.com/siegebell>
 
-declare namespace PEG {
-  function parse(input: string): any;
+/** Interface's that describe the abstract syntax tree used by Peggy. */
```
Comment (phpnode) [inline lib/peg.d.ts:3]:
```suggestion
/** Interfaces that describe the abstract syntax tree used by Peggy. */
```

**Review by phpnode — APPROVED** [review]
LGTM

File: lib/peg.d.ts (hunk 1 above)
Comment (Mingun) [reply lib/peg.d.ts]: Done

**Pull request merged**